### PR TITLE
Reduce goroutine resulting from GetIndexInfos

### PR DESCRIPTION
**Pull request by bigsheeper**
issue: https://github.com/milvus-io/milvus/issues/27545

/kind enhancement

No need to pick to master, because indexCoord has been merged to dataCoord.

### CODE REVIEW

**Review by yah01 — APPROVED** [review]
/lgtm

**Comment by xiaofan-luan** [comment]
when did we call GetIndexinfos?

**Comment by bigsheeper** [comment]
> when did we call GetIndexinfos?

To verify whether a segment has been indexed, for example during compaction or GetRecoveryInfo.

File: internal/datacoord/util.go [hunk 1]
```diff
@@ -195,6 +178,34 @@ func FilterInIndexedSegments(handler Handler, indexCoord types.IndexCoord, segme
...
+	})
+	if err != nil {
+		log.Warn("failed to get index of collection",
+			zap.Int64("collection", collectionID),
+			zap.Int64s("segments", segmentIDs),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	if resp.GetStatus().GetErrorCode() != commonpb.ErrorCode_Success {
```
Comment (yah01) [inline internal/datacoord/util.go:196]: use the new error handling way:
```
err := merr.CheckRPCCall(resp, err)
if err != nil {
// ...
}
```

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/datacoord/util.go (hunk 1 above)
Comment (bigsheeper) [reply internal/datacoord/util.go:196]: this is 2.2.x branch, so no merr

**Review by yah01 — COMMENTED** [review]
/lgtm

File: internal/datacoord/util.go [hunk 2]
```diff
@@ -195,6 +178,34 @@ func FilterInIndexedSegments(handler Handler, indexCoord types.IndexCoord, segme
...
 
+func getIndexInfos(indexCoord types.IndexCoord, collectionID int64, segmentIDs []int64) (map[int64]*indexpb.SegmentInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // TODO: make it configurable
+	defer cancel()
+	resp, err := indexCoord.GetIndexInfos(ctx, &indexpb.GetIndexInfoRequest{
+		CollectionID: collectionID,
+		SegmentIDs:   segmentIDs,
+	})
+	if err != nil {
+		log.Warn("failed to get index of collection",
```
Comment (czs007) [inline internal/datacoord/util.go:189]: better to use log.Info

File: internal/datacoord/util.go [hunk 3]
```diff
@@ -195,6 +178,34 @@ func FilterInIndexedSegments(handler Handler, indexCoord types.IndexCoord, segme
...
+			zap.Int64("collection", collectionID),
+			zap.Int64s("segments", segmentIDs),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	if resp.GetStatus().GetErrorCode() != commonpb.ErrorCode_Success {
+		log.Warn("indexcoord return error when get index",
```
Comment (czs007) [inline internal/datacoord/util.go:197]: better to use Info level

**Pull request merged**